File: search_in_2d_array.cpp
/*
 *                  Search in a 2D Array
 *
 * Time complexity : O(log [nxm])
 * Space complexity : O(1)
 * Did this code successfully run on Leetcode : Yes
 * Any problem you faced while coding this : -
 * Your code here along with comments explaining your approach in three sentences only
 *          - Using binary search
 */
#include<iostream>
#include<vector>

using namespace std;

class Solution {
    public:

    bool searchMatrix(vector<std::vector<int> > matrix, int target) {

        int rows = matrix.size();
        int cols = matrix[0].size();

        cout << "Number of rows : " << rows << endl;
        cout << "Number of cols : " << cols << endl;

        int low = 0;

        int high = rows*cols - 1; // Flattened the sorted array into 1-D array and compared the values.
        
        while (low <= high) {
            int mid = low+high/2;

            if(matrix[mid/cols][mid%cols] == target) return true;

            else if(matrix[mid/cols][mid%cols] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return false;

    }
};

int main() {
    std::vector<std::vector<int> > matrix;
    bool contains = false;

    Solution* obj = new Solution();

    // Initialize each row as a vector
    std::vector<int> row1;
    row1.push_back(1);
    row1.push_back(3);
    row1.push_back(5);
    row1.push_back(7);

    std::vector<int> row2;
    row2.push_back(10);
    row2.push_back(11);
    row2.push_back(16);
    row2.push_back(20);

    std::vector<int> row3;
    row3.push_back(23);
    row3.push_back(30);
    row3.push_back(34);
    row3.push_back(60);

    // Add rows to the matrix
    matrix.push_back(row1);
    matrix.push_back(row2);
    matrix.push_back(row3);

    contains = obj->searchMatrix(matrix, 3);

    cout << contains;

    return 0;
}

File: search-in-2d-array.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  let m = matrix.length;
  let n = matrix[0].length;
  let low = 0;
  let high = m * n - 1;

  while (low <= high) {
    let mid = Math.floor(low + (high - low) / 2);
    let row = Math.floor(mid / n);
    let column = Math.floor(mid % n);

    if (matrix[row][column] == target) {
      return true;
    } else if (matrix[row][column] > target) {
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  return false;
};

var searchMatrixV2 = function (matrix, target) {
  let row = find_row_index(matrix, target);
  console.log("Row " + row);
  if (row === -1) return false;
  let col = find_col_index(matrix[row], target);
  console.log("Col " + row);
  if (col === -1) return false;
  return matrix[row][col] == target;
};

var find_row_index = function (array, target) {
  let left = 0;
  let right = array.length - 1;

  while (left <= right) {
    let mid = Math.floor(left + (right - left) / 2);
    let high = array[mid][array[mid].length - 1];
    let low = array[mid][0];
    if (low <= target && target <= high) {
      return mid;
    } else if (high >= target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }
  return -1;
};

var find_col_index = function (array, target) {
  let left = 0;
  let right = array.length;
  while (left <= right) {
    let mid = Math.floor(left + (right - left) / 2);
    let element = array[mid];
    if (element == target) {
      return mid;
    } else if (element >= target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }
  return -1;
};

let result = searchMatrixV2(
  [
    [1, 3, 5, 7],
    [10, 11, 16, 20],
    [23, 30, 34, 60],
  ],
  3
);
console.log("Result " + result);
